### PR TITLE
Disable AGG(IF()) rewrite for some corner cases

### PR DESCRIPTION
**Pull request by yuanzhanhku**
The rewrite could change the behavior of certain aggregations such as
set_agg(). See the added tests for example. To be safe, we disable it
for all aggregation functions returning arrays.

Test plan
Added query tests in TestFilteredAggregations.java.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/arrayagg/SetAggregationFunction.java [hunk 1]
```diff
@@ -83,6 +83,12 @@ public InternalAggregationFunction specialize(BoundVariables boundVariables, int
         return generateAggregation(type, outputType);
     }
 
+    @Override
+    public boolean isCalledOnNullInput()
+    {
+        return true;
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/arrayagg/SetAggregationFunction.java:89]: We might have the same issue with SET_UNION as well. I mean it maybe missing this override.

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java [hunk 2]
```diff
@@ -124,6 +124,17 @@ public void testGroupAll()
...
 
+    @Test
+    public void testSetAggWithNulls()
+    {
+        assertions.assertQuery(
+                "SELECT y, set_agg(y) FILTER (WHERE x = 1) FROM (SELECT 1 x, 2 y UNION ALL SELECT NULL x, 20 y UNION ALL SELECT 1 x, NULL y) GROUP BY y ORDER BY y",
+                "VALUES (INTEGER '2', ARRAY[INTEGER '2']), (INTEGER '20', CAST(NULL AS ARRAY<INTEGER>)), (CAST(NULL AS INTEGER), ARRAY[CAST(NULL AS INTEGER)])");
+        assertions.assertQuery(
+                "SELECT y, set_agg(IF(x = 1,y)) FROM (SELECT 1 x, 2 y UNION ALL SELECT NULL x, 20 y UNION ALL SELECT 1 x, NULL y) GROUP BY y ORDER BY y",
+                "VALUES (INTEGER '2', ARRAY[INTEGER '2']), (INTEGER '20', ARRAY[CAST(NULL AS INTEGER)]), (CAST(NULL AS INTEGER), ARRAY[CAST(NULL AS INTEGER)])");
```
Comment (kaikalur) [inline presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java:135]: Add a couple of more tests for weird aggs like approx_set, set_union, map_union and map_union_sum.

**Review by kaikalur — CHANGES_REQUESTED** [review]

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/arrayagg/SetAggregationFunction.java (hunk 1 above)
Comment (yuanzhanhku) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/arrayagg/SetAggregationFunction.java:89]: Looks like set_union ignores NULL arrays:
presto:deltoid3> SELECT set_union(e) FROM (VALUES ARRAY[1], CAST(NULL AS ARRAY<BIGINT>)) AS t(e);
 _col0
-------
 [1]
(1 row)

It does preserve NULL array element. But in this case, its input is the array which is not NULL. 
So looks like returning false for isCalledOnNullInput() is correct for set_union?

Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/arrayagg/SetAggregationFunction.java:89]: It has the same issue as set_agg:
```
presto:di> select set_union(if(y > 1, x)) from (select array[1] x, 1 y union all select null, 1 y);
 _col0
-------
 []
(1 row)

Query 20210805_194306_40996_sptmi, FINISHED, 1 node
Splits: 15 total, 15 done (100.00%)
0:00 [0 rows, 0B] [0 rows/s, 0B/s]

presto:di> select set_union(x) filter (where y > 1) from (select array[1] x, 1 y union all select null, 1 y);
 _col0
-------
 NULL
(1 row)
```

Comment (yuanzhanhku) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/arrayagg/SetAggregationFunction.java:89]: Added override function. Thanks for catching this and figuring out the repro example.

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java (hunk 2 above)
Comment (yuanzhanhku) [reply presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java:135]: Done.

**Review by kaikalur — APPROVED** [review]

**Review by highker — APPROVED** [review]

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java [hunk 3]
```diff
@@ -124,6 +124,55 @@ public void testGroupAll()
...
+                "VALUES (CAST (NULL AS MAP<INTEGER, INTEGER>))");
+    }
+
+    @Test
+    public void testMapUnionSum()
+    {
+        assertions.assertQuery("SELECT map_union_sum(x) FILTER (WHERE y > 1) FROM (SELECT MAP(ARRAY[1], ARRAY[1]) x, 1 y UNION ALL SELECT NULL x, 1 y)",
+                "VALUES (CAST (NULL AS MAP<INTEGER, INTEGER>))");
+        assertions.assertQuery("SELECT map_union_sum(IF(y > 1, x)) FROM (SELECT MAP(ARRAY[1], ARRAY[1]) x, 1 y UNION ALL SELECT NULL x, 1 y)",
+                "VALUES (CAST (NULL AS MAP<INTEGER, INTEGER>))");
```
Comment (highker) [inline presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java:173]: leave the first line of params empty

Comment (yuanzhanhku) [reply presto-main/src/test/java/com/facebook/presto/sql/query/TestFilteredAggregations.java]: Done.

**Pull request merged**